### DISCUSSION TITLE
Importing InitFragment class and other types for Dependency within a plugin

### DISCUSSION THREAD

**Discussion by steverep** [webpack/webpack#18230] [Q&A] [answered]
I'm trying to include a dependency similar to `ProvidedDependency` within a plugin, but I need to alter the stage so I need to rewrite the template.  Getting started I'm stumbling on:
1. How to import the `InitFragment` class (or is it not exposed?)
2. How to import the `DependencyTemplateContext` type, which looks to be a prop of `DependencyTemplate` but I cannot find that either.

**Comment by alexander-akait** [comment, marked as answer]
> How to import the InitFragment class (or is it not exposed?)

Feel free to send a PR and expose it :+1: because we already export `Dependency`, so it can be useful

> How to import the DependencyTemplateContext type, which looks to be a prop of DependencyTemplate but I cannot find that either.

You can get it using `dependencies.ModuleDependency.Template`

**Reply by steverep** [reply]
> Feel free to send a PR and expose it 👍 because we already export Dependency, so it can be useful

I can do that but I'm not sure that's the only thing needed?  At first glance the serializing utilities would be needed as well?  I'm still trying to understand what that does and if I need to implement it.new to me

**Reply by steverep** [reply]
See #18232, but it looks like the `makeSerializable` utility should also be exported?  Again it's unclear to me how this functions without digging a lot deeper.

**Reply by alexander-akait** [reply]
We have an API for this, you can find an example here - https://github.com/webpack-contrib/mini-css-extract-plugin/blob/master/src/index.js#L354

**Reply by steverep** [reply]
Off the original topic, but is there a way to get a reference to the parent module from inside a parser hook (i.e. find the caller for `parser.state.module`)?  Specifically, I'm wondering if instead of  tapping `finishModules` like:

https://github.com/webpack/webpack/blob/fb7c0c637ad8930dff05624f87aa7ba06bf4fe92/lib/async-modules/InferAsyncModulesPlugin.js#L22

Would it be equivalent to get the same list of async modules by:
1. Flagging TLA modules like it already does
2. Flagging the parent/caller while still parsing

This would save a lot of trouble, but not sure if I'm missing something or there's no reliable way to do it.

**Reply by steverep** [reply]
@alexander-akait ignore my last.  I figured out how to tap `afterPlugins` to let all async modules be set first, then tap `finishModules` to add my dependencies.  However, I'm running into something that seems like a bug to me (or I'm just doing it very wrong).

My hook looks something like this:

```ts
    compiler.hooks.compilation.tap(
      PLUGIN_NAME,
      (compilation, { normalModuleFactory }) => {
        // Not sure what this does...
        compilation.dependencyFactories.set(
          TransformAsyncDependency,
          normalModuleFactory,
        );
        compilation.dependencyTemplates.set(
          TransformAsyncDependency,
          new TransformAsyncDependency.Template(),
        );
        compilation.hooks.finishModules.tap(PLUGIN_NAME, (modules) => {
          for (const module of modules) {
            if (compilation.moduleGraph.isAsync(module)) {
              // module.addDependency()  need this?
              compilation.moduleGraph.updateModule(
                new TransformAsyncDependency(
                  "@babel/runtime/helpers/asyncToGenerator",
                  "_asyncToGenerator",
                  ["default"],
                ),
                module,
              );
            }
          }
        });
      },
    );
```

The `updateModule` method immediately calls `getConnection`, which returns `undefined` and then throws because the next line tries to access `connection.module`:

https://github.com/webpack/webpack/blob/3e3dfd781b9f1b6a64424cdf79f6187db797eb0e/lib/ModuleGraph.js#L244-L249

If I add the dependency via `module.addDependency` like `ProvidePlugin` instead of (or in addition to) `updateModule`, the same error occurs because the first thing the template does is call `getConnection`.

So what am I doing wrong?  Or is this a bug and `updateModule`  should be handling the missed connection in order to add it?

**Reply by alexander-akait** [reply]
Why do not use `parser.hooks`? Do you want to inject `@babel/runtime/helpers/asyncToGenerator` only for async modules? You need to `addDependency` when parser works

**Reply by steverep** [reply]
Yes, only for async modules.  For modules with TLA, it's easy enough to tap `parser.hooks.topLevelAwait`, then `parser.state.module.addDependency()`.  But the parent/issuer is also async, which by then is already parsed.  So is there any method in the API to add dependencies to a module that's already been parsed?

**Reply by alexander-akait** [reply]
There is a `buildModule` hook, you can try it just using `addDependency`

**Reply by steverep** [reply]
But the parent/issuer module would have already been built by the time the TLA is found by the parser, so wouldn't that be too late as well?

**Reply by alexander-akait** [reply]
hm, `buildModule` runs before the building module :smile: Can you discribe me what you try to solve to get a full context, maybe there are more simple solutions

**Reply by steverep** [reply]
I managed to figure it out by rebuilding the modules after adding the dependencies.  Async modules consist of those with TLA plus any that import them, and I want to add dependencies to all of them.  So the issue is that you need to find the TLA through parsing first, and by the time you do that the module that imported it has already been built.

Here's what I came up with that seems to work:

https://github.com/steverep/transform-async-modules-webpack-plugin/blob/747c1b1d095acfb9340d9295a25b7d8d76955a9f/src/index.cts#L107-L177

If you have any suggestions to improve or simplify, please do tell.  I'm mostly waiting on a release that exports `InitFragment` from #18232 in order to complete the work.

**Reply by alexander-akait** [reply]
Code looks good, maybe you can implement caching based on hashes to improve perf

**Reply by steverep** [reply]
Yeah that might be the next step... just need to get it fully functional first.  I'm trying to avoid needing to add an `InitFragment`, but it's cumbersome and ugly.

**Reply by steverep** [reply]
Actually, I should just ask you if there's a better way.  The process using an `InitFragment` would look like:

1. Each helper function Babel wants to import from `@babel/runtime` is created as a dependency.
2. The dependency template works just like `ProvidePlugin`, except it creates an `InitFragment` with a stage that puts it outside the async wrapper boundary.  It injects a "temporary" identifier like `var _temp_asyncToGenerator = __webpack_require__(...);`.
3. Babel transforms it, injecting any helpers like `import _asyncToGenerator from "..."` (the identifier it uses is guaranteed to be unique and will be used in the transformed code).
4. My Webpack plugin uses a little extra custom Babel plugin to essentially just swap the import for the `__webpack_require__` using the correct identifier.

So the question is:  Can I alter step 2 so that the template just saves the `__webpack_require__` expression string somewhere, such that I can retrieve it from the `renderModuleContent` hook?  I'm not sure how (or if it's cleanly possible) to do that safely since the expression is a function of quite a few things.

**Reply by alexander-akait** [reply]
> So the question is: Can I alter step 2 so that the template just saves the __webpack_require__ expression string somewhere, such that I can retrieve it from the renderModuleContent hook? I'm not sure how (or if it's cleanly possible) to do that safely since the expression is a function of quite a few things.

I am afraid no, you don't use loader for this, because you want to transform runtime webpack code, right?

**Reply by steverep** [reply]
Correct, I'm transpiling the async function introduced by Webpack when handling top level awaits.  I looked into getting around an `InitFragment` and determined it was far more elegant just to use it.  Here's a semi-final product: https://github.com/steverep/transform-async-modules-webpack-plugin/pull/32.


When testing on a large project though, I'm running into a snag caused by Babel also needing to transpile a destructured assignment used by Webpack.  It looks like checking whether it is supported is marked as a TODO:

https://github.com/webpack/webpack/blob/main/lib/async-modules/AwaitDependenciesInitFragment.js

**Reply by steverep** [reply]
Was able to get around the problem of destructuring with a Babel setting.  Now I'm down to some Webpack errors when optimizing that seem to self-blame back to Webpack:


```
ERROR in ./src/common/datetime/absolute_time.ts
  RuntimeTemplate.importStatement(): Module /workspaces/frontend/node_modules/babel-loader/lib/index.js??ruleSet[1].rules[0].use!/workspaces/frontend/node_modules/date-fns/esm/index.js has no id assigned.
  This should not happen.
  It's in these chunks: none (If module is in no chunk this indicates a bug in some chunk/module optimization logic)
  Module has these incoming connections: 
 ...
```

**Reply by steverep** [reply]
The errors go away with `optimization.usedExports = false`, but I have no idea what the cause is.  I get a few of those errors all with the same module without an ID, and all in referencing modules that are not async but do import async modules.

Could this somehow be cause by a flaw in my dependency subclass?

**Reply by steverep** [reply]
Just following up to say I figured it out and released a new version.  Running `compilation.processModuleDependencies` a 2nd time was corrupting things.  I ended up just taking apart the method and stripping it down to a couple lines for adding dependencies after the fact:

https://github.com/steverep/transform-async-modules-webpack-plugin/blob/debb91e7cc0da76c7c9d36312388a7eaeeb7a982/src/index.cts#L166-L208

Anyway, thanks for your guidance.  Here's a PR for Home Assistant to demo why this was worth it: https://github.com/home-assistant/frontend/pull/20543.

**Reply by alexander-akait** [reply]
Sorry for delay, a lot of issue, glad to hear that you found a solution :heart:
